File: c-algorithm/test/test-binomial-heap.c
// 识别到的包含指令
#include <stdlib.h>
#include <assert.h>
#include "alloc-testing.h"
#include "framework.h"

// 识别到的全局变量
_Atomic size_t num_assert = 0;

// 识别到的宏定义
#define assert(expr)                                                           \
  num_assert += 1;                                                             \
  ((void)sizeof((expr) ? 1 : 0), __extension__({                               \
     if (expr)                                                                 \
       ; /* empty */

// 识别到的包含指令
#include "binomial-heap.h"
#include "compare-int.h"

// 识别到的宏定义
#define NUM_TEST_VALUES 10000

// 识别到的全局变量
int test_array[NUM_TEST_VALUES];

// 识别到的函数定义
// 函数: test_binomial_heap_new_free (line 45)
void test_binomial_heap_new_free(void)
{
	BinomialHeap *heap;
	int i;

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		heap = binomial_heap_new(BINOMIAL_HEAP_TYPE_MIN, int_compare);
		binomial_heap_free(heap);
	}

	/* Test for out of memory */

	alloc_test_set_limit(0);

	assert(binomial_heap_new(BINOMIAL_HEAP_TYPE_MIN, int_compare) == NULL);
}

// 函数: test_binomial_heap_insert (line 62)
void test_binomial_heap_insert(void)
{
	BinomialHeap *heap;
	int i;

	heap = binomial_heap_new(BINOMIAL_HEAP_TYPE_MIN, int_compare);

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		test_array[i] = i;
		assert(binomial_heap_insert(heap, &test_array[i]) != 0);
	}
	assert(binomial_heap_num_entries(heap) == NUM_TEST_VALUES);

	/* Test for out of memory */

	alloc_test_set_limit(0);
	assert(binomial_heap_insert(heap, &i) == 0);

	binomial_heap_free(heap);
}

// 函数: test_min_heap (line 83)
void test_min_heap(void)
{
	BinomialHeap *heap;
	int *val;
	int i;

	heap = binomial_heap_new(BINOMIAL_HEAP_TYPE_MIN, int_compare);

	/* Push a load of values onto the heap */

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		test_array[i] = i;
		assert(binomial_heap_insert(heap, &test_array[i]) != 0);
	}

	/* Pop values off the heap and check they are in order */

	i = -1;
	while (binomial_heap_num_entries(heap) > 0) {
		val = (int *) binomial_heap_pop(heap);

		assert(*val == i + 1);
		i = *val;
	}

	/* Test pop on an empty heap */

	val = (int *) binomial_heap_pop(heap);
	assert(val == NULL);

	binomial_heap_free(heap);
}

// 函数: test_max_heap (line 116)
void test_max_heap(void)
{
	BinomialHeap *heap;
	int *val;
	int i;

	heap = binomial_heap_new(BINOMIAL_HEAP_TYPE_MAX, int_compare);

	/* Push a load of values onto the heap */

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		test_array[i] = i;
		assert(binomial_heap_insert(heap, &test_array[i]) != 0);
	}

	/* Pop values off the heap and check they are in order */

	i = NUM_TEST_VALUES;
	while (binomial_heap_num_entries(heap) > 0) {
		val = (int *) binomial_heap_pop(heap);

		assert(*val == i - 1);
		i = *val;
	}

	/* Test pop on an empty heap */

	val = (int *) binomial_heap_pop(heap);
	assert(val == NULL);

	binomial_heap_free(heap);
}

// 识别到的宏定义
#define TEST_VALUE (NUM_TEST_VALUES / 2)

// 识别到的函数定义
// 函数: generate_heap (line 151)
static BinomialHeap *generate_heap(void)
{
	BinomialHeap *heap;
	int i;

	heap = binomial_heap_new(BINOMIAL_HEAP_TYPE_MIN, int_compare);

	/* Push a load of values onto the heap */

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		test_array[i] = i;
		if (i != TEST_VALUE) {
			assert(binomial_heap_insert(heap,
			                            &test_array[i]) != 0);
		}
	}

	return heap;
}

// 函数: verify_heap (line 174)
static void verify_heap(BinomialHeap *heap)
{
	unsigned int num_vals;
	int *val;
	int i;

	num_vals = binomial_heap_num_entries(heap);
	assert(num_vals == NUM_TEST_VALUES - 1);

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		if (i == TEST_VALUE) {
			continue;
		}

		/* Pop off the next value and check it */

		val = binomial_heap_pop(heap);
		assert(*val == i);

		/* Decrement num values counter */

		--num_vals;
		assert(binomial_heap_num_entries(heap) == num_vals);
	}
}

// 函数: test_insert_out_of_memory (line 202)
static void test_insert_out_of_memory(void)
{
	BinomialHeap *heap;
	int i;

	/* There are various memory allocations performed during the insert;
	 * probe at different limit levels to catch them all. */

	for (i=0; i<6; ++i) {
		heap = generate_heap();

		/* Insert should fail */

		alloc_test_set_limit(i);
		test_array[TEST_VALUE] = TEST_VALUE;
		assert(binomial_heap_insert(heap,
		                            &test_array[TEST_VALUE]) == 0);
		alloc_test_set_limit(-1);

		/* Check that the heap is unharmed */

		verify_heap(heap);

		binomial_heap_free(heap);
	}
}

// 函数: test_pop_out_of_memory (line 231)
void test_pop_out_of_memory(void)
{
	BinomialHeap *heap;
	int i;

	/* There are various memory allocations performed as part of the merge
	 * done during the pop.  Probe at different limit levels to catch them
	 * all. */

	for (i=0; i<6; ++i) {
		heap = generate_heap();

		/* Pop should fail */

		alloc_test_set_limit(i);
		assert(binomial_heap_pop(heap) == NULL);
		alloc_test_set_limit(-1);

		/* Check the heap is unharmed */

		binomial_heap_free(heap);
	}
}

// 识别到的全局变量
static UnitTestFunction tests[] = {
	test_binomial_heap_new_free,
	test_binomial_heap_insert,
	test_min_heap,
	test_max_heap,
	test_insert_out_of_memory,
	test_pop_out_of_memory,
	NULL
};

// 识别到的函数定义
// 函数: main (line 265)
int main(int argc, char *argv[])
{
	run_tests(tests);
  printf("num_assert: %lu\n", num_assert);
	return 0;
}

File: comparisons/MinsC2Rust/c_algorithm/Output/test_project/tests/test_binomial_heap.rs
use test_project::binomial_heap::{
    binomial_heap_cmp, binomial_heap_free, binomial_heap_insert, binomial_heap_merge,
    binomial_heap_merge_undo, binomial_heap_new, binomial_heap_num_entries, binomial_heap_pop,
    binomial_tree_merge, binomial_tree_ref, binomial_tree_unref, BinomialHeap, BinomialHeapType,
    BinomialTree, NUM_TEST_VALUES, TEST_VALUE,
};
pub fn generate_heap() -> Option<BinomialHeap<i32>> {
    let mut heap = binomial_heap_new(BinomialHeapType::Min, |a: &i32, b: &i32| a.cmp(b) as i32)?;

    let mut test_array = vec![0; NUM_TEST_VALUES];

    for i in 0..NUM_TEST_VALUES {
        test_array[i] = i as i32;
        if i != TEST_VALUE {
            assert!(binomial_heap_insert(&mut heap, test_array[i]));
        }
    }

    Some(heap)
}

pub fn verify_heap<T: Clone + PartialEq + std::fmt::Debug>(
    heap: &mut BinomialHeap<T>,
    test_values: &[T],
    test_value_index: usize,
) {
    let mut num_vals = binomial_heap_num_entries(heap);
    assert_eq!(num_vals, test_values.len() - 1);

    for i in 0..test_values.len() {
        if i == test_value_index {
            continue;
        }

        let val = binomial_heap_pop(heap);
        assert_eq!(val.as_ref(), Some(&test_values[i]));

        num_vals -= 1;
        assert_eq!(binomial_heap_num_entries(heap), num_vals);
    }
}

#[test]
pub fn test_min_heap() {
    let mut heap = binomial_heap_new(BinomialHeapType::Min, int_compare).unwrap();
    let mut test_array: Vec<i32> = (0..NUM_TEST_VALUES as i32).collect();

    for i in 0..NUM_TEST_VALUES {
        assert!(binomial_heap_insert(&mut heap, test_array[i]));
    }

    let mut i = -1;
    while binomial_heap_num_entries(&heap) > 0 {
        let val = binomial_heap_pop(&mut heap).unwrap();
        assert_eq!(val, i + 1);
        i = val;
    }

    let val = binomial_heap_pop(&mut heap);
    assert!(val.is_none());

    binomial_heap_free(heap);
}

#[test]
pub fn test_insert_out_of_memory() {
    let mut heap: Option<BinomialHeap<i32>>;

    for _ in 0..6 {
        heap = generate_heap();

        let mut test_array = vec![0; NUM_TEST_VALUES];
        for i in 0..NUM_TEST_VALUES {
            test_array[i] = i as i32;
        }

        test_array[TEST_VALUE] = TEST_VALUE as i32;
        assert!(!binomial_heap_insert(
            heap.as_mut().unwrap(),
            test_array[TEST_VALUE]
        ));

        verify_heap(heap.as_mut().unwrap(), &test_array, TEST_VALUE);

        binomial_heap_free(heap.unwrap());
    }
}

#[test]
pub fn test_pop_out_of_memory() {
    for _ in 0..6 {
        let mut heap = generate_heap().unwrap();

        // Pop should fail
        assert!(binomial_heap_pop(&mut heap).is_none());

        // Check the heap is unharmed
        binomial_heap_free(heap);
    }
}

#[test]
pub fn test_binomial_heap_new_free() {
    for _ in 0..NUM_TEST_VALUES {
        let heap = binomial_heap_new(BinomialHeapType::Min, |a: &i32, b: &i32| match a.cmp(b) {
            std::cmp::Ordering::Less => -1,
            std::cmp::Ordering::Equal => 0,
            std::cmp::Ordering::Greater => 1,
        });
        if let Some(heap) = heap {
            binomial_heap_free(heap);
        }
    }
}

#[test]
pub fn test_max_heap() {
    let mut heap =
        binomial_heap_new(BinomialHeapType::Max, |a: &i32, b: &i32| a.cmp(b) as i32).unwrap();
    let mut test_array: Vec<i32> = (0..NUM_TEST_VALUES as i32).collect();

    for i in 0..NUM_TEST_VALUES {
        assert!(binomial_heap_insert(&mut heap, test_array[i]));
    }

    let mut i = NUM_TEST_VALUES as i32;
    while binomial_heap_num_entries(&heap) > 0 {
        let val = binomial_heap_pop(&mut heap).unwrap();
        assert_eq!(val, i - 1);
        i = val;
    }

    let val = binomial_heap_pop(&mut heap);
    assert!(val.is_none());

    binomial_heap_free(heap);
}

#[test]
pub fn test_binomial_heap_insert() {
    let mut heap = binomial_heap_new(BinomialHeapType::Min, int_compare).unwrap();
    let test_array: Vec<i32> = (0..NUM_TEST_VALUES as i32).collect();

    for i in 0..NUM_TEST_VALUES {
        assert!(binomial_heap_insert(&mut heap, test_array[i]));
    }
    assert_eq!(binomial_heap_num_entries(&heap), NUM_TEST_VALUES);

    binomial_heap_free(heap);
}

pub fn int_compare(a: &i32, b: &i32) -> i32 {
    if a < b {
        -1
    } else if a > b {
        1
    } else {
        0
    }
}
